### PR TITLE
[MSHARED-1007] Add MavenHome and MavenExecutable options to InvocationRequest

### PR DESCRIPTION
**Pull request by slawekjaranowski**

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/invoker/InvocationRequest.java [hunk 1]
```diff
@@ -763,4 +763,45 @@ public static ReactorFailureBehavior valueOfByLongOption( String longOption )
...
+     *         Maven installation.
+     *
+     * @since 3.1.1
+     */
+    File getMavenHome();
+
+    /**
+     * Sets the path to the base directory of the Maven installation used to invoke Maven. This parameter may be left
+     * unspecified to use the default Maven installation which will be discovered by evaluating the system property
+     * <code>maven.home</code> and the environment variable <code>M2_HOME</code>.
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/invoker/InvocationRequest.java:779]: The `M2_HOME`  variable is long dead. Don't use it.

File: src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java [hunk 2]
```diff
@@ -599,46 +578,61 @@ protected void setThreads( InvocationRequest request, Commandline cli )
...
     {
+        if ( request.getMavenHome() != null )
+        {
+            mavenHome = request.getMavenHome();
+        }
+
         if ( mavenHome == null )
         {
             String mavenHomeProperty = System.getProperty( "maven.home" );
+            if ( mavenHomeProperty == null && getSystemEnvVars().getProperty( "M2_HOME" ) != null )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java:591]: Drop this.

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by slawekjaranowski** [comment]
There is issue for removing M2_HOME - https://issues.apache.org/jira/browse/MSHARED-577

If it is only this remarks I will merge this PR and in separate change remove all occurrence of M2_HOME
ok? - @michael-o

File: src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java [hunk 3]
```diff
@@ -599,46 +578,61 @@ protected void setThreads( InvocationRequest request, Commandline cli )
...
-                    {
-                        throw new IllegalStateException( "${maven.home} is not specified as a directory: '"
-                            + mavenHomeProperty + "'." );
-                    }
-                }
             }
+        }
 
-            if ( ( mavenHome == null ) && ( getSystemEnvVars().getProperty( "M2_HOME" ) != null ) )
+        if ( mavenHome != null && !mavenHome.isDirectory() )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java:602]: This confuses me. Do you assume that someone will pass the executable to `mavenHome`?

Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java:602]: This code was moved .. someone do it in past
I didn't want change too much.

**Comment by michael-o** [comment]
> There is issue for removing M2_HOME - https://issues.apache.org/jira/browse/MSHARED-577
> 
> If it is only this remarks I will merge this PR and in separate change remove all occurrence of M2_HOME ok? - @michael-o

Go ahead an then remove `M2_HOME` as part of that other ticket.

**Comment by jira-importer** [comment]
Resolve #141

**Pull request merged**